### ISSUE TITLE
Architecture: Unit tests are non-hermetic

### ISSUE THREAD

**Issue by audiodude** [openzim/mwoffliner#2057]
Many (most?) of the unit tests actually make network requests to various MediaWiki projects like English Wikipedia. This leads them to be inherently flaky, because changes to data (not the API, just data) in these wikis can cause the tests to fail. See [here](https://medium.com/geekculture/hermetic-unit-testing-8c49276e3acd) for a simple and direct overview of the problem and the value of hermetic tests.

It is understood that mwoffliner needs to be tested against live wikis in order to prevent regressions. However, tests should be segregated so that any test which requires a live network connection is moved into e2e tests.

We should be able to test specific behaviors like "This utility method returns normalized titles" without downloading entire wiki pages.

Fixing this would also likely lead to architectural improvements, because we would be forced to decouple various modules more and use [dependency injection](https://www.jamesshore.com/v2/blog/2006/dependency-injection-demystified).

It would also make the unit tests more speedy, because the network requests would not be necessary.

E2E tests could still be run in CI and on demand, when necessary.

**Comment by kelson42** [comment]
I can only agree with you, but not that easy to mockup MediWiki with different configurations. That said we have an easy way to deply MediaWiki with https://github.com/offspot/mediawiki-docker

**Comment by audiodude** [comment]
Yes, one of the ways to make the unit tests hermetic would be to have our own MediaWiki installations.

But I was thinking more along the lines of https://github.com/openzim/mwoffliner/pull/2043/files

Where the unit test just tests one aspect of the code and does not connect to any MediaWiki installation at all. I believe unit tests should test individual aspects and functions, instead of "Download this whole list of articles and then test if this one portion of it worked", which seems to be the current architecture.

**Comment by rgaudin** [comment]
## GSoC guidance

MWoffliner’s test suite currently relies heavily on live HTTP requests to Wikipedia and other MediaWiki sites, resulting in 30 minutes test runs that are flaky, resource-intensive, and unsuitable for CI/CD pipelines. This project aims to rewrite the test infrastructure to be hermetic—using mocked responses, local test fixtures, and in-memory MediaWiki instances where needed.

Deliverables: Contributors will identify all external dependencies, create comprehensive mock data representing various MediaWiki configurations and edge cases, refactor tests to use dependency injection for HTTP clients, and establish patterns for future hermetic testing. The goal is reducing test runtime to under 10 minutes while improving coverage and reliability. By the end, the tests should be as hermetic as possible. A stretch goal is to separate any remaining non-hermetic tests (which need to make HTTP calls for a technical reason) to a separate e2e test target.

Skills required: Strong testing experience (unit, integration, mocking frameworks), TypeScript/Node.js proficiency, understanding of HTTP mocking techniques (nock, MSW, or similar), experience with CI/CD systems. Familiarity with MediaWiki structure and test architecture design patterns is a plus: use the time before applications are opened to familiarize yourself with the tool.

Project Length: Medium (150-175 hrs in total)

**Comment by benoit74** [comment]
Thank you

**Comment by nishantharkut** [comment]
Hey, 
Following up on the GSoC guidance from @rgaudin and the approach @audiodude outlined with PR #2043, I spent time reading the test suite to map out the scope.     
                                                                                                                                                       
The biggest single target is `setupScrapeClasses()` in **test/util.ts** — it calls **getMwMetaData()**, **hasCoordinates(),** and **hasModuleApi()** live on every invocation. That's 5 unit test files depending on it: `saveArticles.test.ts`, `urlRewriting.test.ts`, `media.treatment.test.ts`, `article.treatment.test.ts`, and `renderer.builder.test.ts`. 
Replacing it with a fixture-based version would take all 5 offline at once.                                           

Another 4 files call MediaWiki APIs directly without going through **setupScrapeClasses**: `mwApi.test.ts`, `mwApiCapabilities.test.ts`,    `article.renderer.test.ts`, and `downloader.test.ts`. These would need individual mocking - nock or MSW as suggested.
                                                                                                                                                       
There's also convertWikicodeToHtml() in test/util.ts which POSTs to Wikipedia's REST API, used by media.treatment.test.ts.                      

**Existing patterns to build on**: `downloader.retry.test.ts` already mocks `request()` with `jest.fn()` and runs fully offline, and **mockRedis.ts** provides an in-memory Redis mock. The TODOs in **saveArticles.test.ts** and **article.treatment.test.ts** ("use proper spied like sinon.js") suggest this direction was already intended.                                                                                                                                    
                  
Total: 9 non-hermetic unit test files vs ~19 hermetic ones. Each file can be migrated as an independent PR.                                        
  
I'd like to work on this for GSoC.  I'm submitting a small code quality PR shortly to get familiar with the contribution workflow.
